### ISSUE TITLE
None of `onCreateCommand`, `updateContentCommand`, and `postCreateCommand` are run on container re-creation

### ISSUE THREAD

**Issue by posita** [microsoft/vscode-remote-release#6336]
The docs for `onCreateCommand`, `updateContentCommand`, and `postCreateCommand` indicate that they are run inside the container upon container creation. Consider the case where `devcontainer.json` builds off of a `Dockerfile` and that underlying `Dockerfile` changes. Currently, when the container is rebuilt, a new container is created, but none of those three commands are executed, which makes them seem kind of useless. There doesn't appear to be a command that is run after the container is recreated. `postStartCommand` runs every time the container starts, which seems like overkill for the suggested use case of performing additional setup (e.g., via `pip`).

**Comment by chrmarti** [comment]
We see this when the container has `~/.vscode-server` or one of its parent folders mounted, such that its content is preserved between containers. Is that the case? Could you append your devcontainer.json?

**Comment by posita** [comment]
> We see this when the container has `~/.vscode-server` or one of its parent folders mounted, such that its content is preserved between containers. Is that the case?

I have a named volume that targets the executing user's home directory, which results in `~/.vscode-server` being preserved between runs. Does that imply that the detection mechanism's implementation depends on the absence or presence of `~/.vscode-server` (or something inside it)?

> Could you append your devcontainer.json?

What does this mean? How would I append it?

**Comment by chrmarti** [comment]
There are marker files inside `~/.vscode-server` keeping track of which commands were run.

**Comment by posita** [comment]
> There are marker files inside `~/.vscode-server` keeping track of which commands were run.

That seems like an implementation bug. There's no requirement stated in [the docs](https://code.visualstudio.com/docs/remote/devcontainerjson-reference#_lifecycle-scripts) that one must avoid persisting the executing user's home outside the container in order for those commands to work. Nor do I think that requirement would be reasonable. In other words, I'm not advocating for updating the docs as a "fix". My guess is the approach has to change.

Where are the marker files, and what is responsible for checking them? (I'm not seeing much [here](https://github.com/microsoft/vscode/search?q=onCreateCommand), so I'm assuming I'm looking in the wrong place?)

**Comment by chrmarti** [comment]
Adding an option for where the `.vscode-server` should be placed would help. That is tracked as https://github.com/microsoft/vscode-remote-release/issues/6079.

The marker files (names starting with `.`) are in `~/.vscode-server/data/Machine`.

I can include a timestamp in the marker files. We already do that for `postStartCommand`. This will make it work as long as the `.vscode-server` folder is not shared between different dev containers.

**Comment by posita** [comment]
@chrmarti, did you close this because you're treating it as a duplicate of #6079? That issue doesn't seem to capture your proposal:

> I can include a timestamp in the marker files. We already do that for `postStartCommand`. This will make it work as long as the `.vscode-server` folder is not shared between different dev containers.

That seems like a reasonable change, regardless of where `.vscode-server` lives. Should I open a new issue to track that work?

**Comment by chrmarti** [comment]
I have closed it with the timestamp fix implemented. That will be in the upcoming pre-release version 0.223.0 of the Remote-Containers extension. (You'll need to switch Remote-Containers to use pre-release versions if you want to get that version.)

**Comment by posita** [comment]
![That was…fast](https://c.tenor.com/vcTA-90I9JUAAAAM/sonny-carisi-carisibot.gif)

Thank you!

![Respect](https://c.tenor.com/-h4Vr0o9Q8oAAAAM/bowing-down-praise.gif)

**Comment by chrmarti** [comment]
Published Remote-Containers 0.223.0 (pre-release) with the fix.

**Comment by posita** [comment]
I can verify this "works", but _**only**_ if you blow away any existing `~/.vscode-server` directory in the `remoteUser`'s long-running, volume-mounted home directory that existed prior to the update. Until I did that, `postCreateCommand` would not run, even for 0.224.1. Only `postStartCommand` would.

Hopefully that makes sense to you, but that may be a source of confusion for those similarly situated who still experience this bug even after updating their extensions.

**Comment by chrmarti** [comment]
@posita You are right, the fix only works if there is no empty marker file present. I will leave it at that unless this turns out to be problem for too many users. Thanks!

**Comment by posita** [comment]
Where is the empty marker and why not create it if it is missing?

**Comment by posita** [comment]
Understood. I updated the original description to prominently call out any sharp edges for those relatively few folks who might stumble across this issue. Feel free to tweak it as you see fit.
